perfkitbenchmarker/linux_benchmarks/iperf_benchmark.py

```diff
@@ -98,63 +128,241 @@ def _RunIperf(sending_vm, receiving_vm, receiving_ip_address, thread_count,
-    # below will tend to overestimate a bit.
-    thread_values = re.findall(r'\[.*\d+\].*\s+(\d+\.?\d*).Mbits/sec', stdout)
-
-    if len(thread_values) != thread_count:
-      raise ValueError('Only %s out of %s iperf threads reported a'
-                       ' throughput value.' %
-                       (len(thread_values), thread_count))
-
-  total_throughput = 0.0
-  for value in thread_values:
-    total_throughput += float(value)
-
-  metadata = {
-      # The meta data defining the environment
-      'receiving_machine_type': receiving_vm.machine_type,
-      'receiving_zone': receiving_vm.zone,
-      'sending_machine_type': sending_vm.machine_type,
-      'sending_thread_count': thread_count,
-      'sending_zone': sending_vm.zone,
-      'runtime_in_seconds': FLAGS.iperf_runtime_in_seconds,
-      'ip_type': ip_type
-  }
-  return sample.Sample('Throughput', total_throughput, 'Mbits/sec', metadata)
+
+  if protocol == TCP:
+
+    iperf_cmd = ('iperf -e --client %s --port %s --format m --time %s -P %s' %
+                 (receiving_ip_address, IPERF_PORT,
+                  FLAGS.iperf_runtime_in_seconds,
+                  thread_count))
+    if FLAGS.iperf_tcp_per_stream_bandwidth:
+      iperf_cmd = iperf_cmd + (' -b %dM' % FLAGS.iperf_tcp_per_stream_bandwidth)
+    # the additional time on top of the iperf runtime is to account for the
+    # time it takes for the iperf process to start and exit
+    timeout_buffer = FLAGS.iperf_timeout or 30 + thread_count
+    stdout, _ = sending_vm.RemoteCommand(iperf_cmd, should_log=True,
+                                         timeout=FLAGS.iperf_runtime_in_seconds + timeout_buffer)
+
+    multi_thread = re.findall(r'\[SUM\]\s+\d+\.\d+-\d+\.\d+\s\w+\s+\d+\s\w+\s+\d+\s\w+\/\w+\s+\d+\/\d+\s+\d+\s+', stdout)
+    window_size = re.findall(r'TCP window size: \d+\.\d+ \S+', stdout)
+    buffer_size_re = re.findall(r'Write buffer size: \d+\.\d+ \S+', stdout)
+    buffer_size = re.findall(r'\d+\.\d+', str(buffer_size_re))
+    buffer_size_num = float(buffer_size[0])
+    # buffer_size_unit = re.findall(r'\d+\.\d+ (\S+)', buffer_size_re[0])
+    window_size_num = (re.findall(r'\d+\.\d+', str(window_size)))
+    window_size_num = float(window_size_num[0])
+    window_size_measurement = re.findall(r'\d+\.\d+ (\S+)', window_size[0])
+    window_size_measurement = window_size_measurement[0]
+
+    # Iperf output is formatted differently when running with multiple threads vs a single thread
+    if multi_thread:
+      # Write and Err
+      write_err = re.findall(r'\d+ Mbits\/sec\s+(\d+\/\d+)', str(multi_thread))
+      write_re = re.findall(r'\d+', str(write_err))
+      write = int(write_re[0])
+      err = int(write_re[1])
+      # Retry
+      retry_re = re.findall(r'\d+ Mbits\/sec\s+ \d+\/\d+\s+(\d+)', str(multi_thread))
+      retry = int(retry_re[0])
+
+      # Cwnd
+      cwnd_rtt = re.findall(r'\d+ Mbits\/sec\s+ \d+\/\d+\s+\d+\s+(-*\d+\w+\-*/\d+\s+\w+)', stdout)
```

Comment:
I would do this all in one go:
```
    match = re.seach(r'\d+ Mbits\/sec\s+ (?P<write>\d+)/(?<err>\d+)\s+(?P<retry>\d+)\s+(?{<cwnd>-*\d+\w+\-*/\d+\s+\w+)', stdout)
   write, err = int(match.group('write')), int(match.group('err'))
   retry = int(match.group('retry'))
   # use match.group('cwnd')
```

Reply:
I was unaware of the named capture groups for regex. This is much better!

Reply:
I rewrote a lot of the regex statements.
It should make them both more readable and compact.